### ISSUE TITLE
Invalid Hash implementation

### ISSUE THREAD

**Issue by orlp** [reem/rust-ordered-float#142]
Currently the `Hash` implementation just assumes that the `FloatCore` type has a mantissa/exponent width like `f64`. This is not necessarily true, one could have a valid custom `FloatCore` type that does not fit these constraints.

Furthermore, the hash implementation is rather slow, because it reconstructs the type from `integer_decode` (<https://rust.godbolt.org/z/hTTqnaso6>). That's 10-15 cycles every float for what should be a no-op - this can be more expensive than the hash function itself!

I propose we add a `StdFloat` sealed trait that is implemented for `f32` and `f64` only, and use that as the bound for the `Hash` implementation. Then the hash implementation can use `.to_bits()`.

Unfortunately, this is a breaking change. However I must reiterate that the current implementation is *invalid* for the `FloatCore` trait bound, it's possible to make a perfectly sane type implementing `FloatCore` which would violate the `Eq`/`Hash` parity when used in `OrderedFloat` due to no fault of the original type.

---

An alternative solution is to feed the sign, mantissa and exponent returned by `f.integer_decode()` separately into the hasher. However this would make the hash implementation slower still.

**Comment by mbrubeck** [comment]
> Can you give an example of a correct `FloatCore` implementation that leads to an invalid `Hash` implementation?

I guess one possibility is that the mantissa is not normalized, so that the number 1 (for example) could be represented in multiple ways like `(1, 0, 1)` and `(2, -1, 1)`.

**Comment by orlp** [comment]
> We throw away the top few bits of the mantissa and exponent, which could be used in some custom implementations. This may increase collisions, but should not cause the Hash implementation to be incorrect.

You are right, extra collisions technically don't make it incorrect. Although I do such consider systematic collisions problematic. So my initial statement was overstated.

> Ah, I guess one possibility is that the mantissa is not normalized, so that the number 2 (for example) could be represented in multiple ways like (1, 1, 1) and (2, 0, 1) and (4, -1, 1).

That said, you are right, this once again does make it *technically* incorrect.

> Ouch, this is unfortunate. I wish we had specialization so we could easily fix this without a breaking change...

So do I... so do I...

> Or if necessary we could implement Hash directly just for `OrderedFloat<f32>`, `OrderedFloat<f64>`, `NotNan<f32>`, and `NotNan<f64>`.

My biggest concern with this is that it breaks generic code of users in a way that's more annoying to fix. At least with a `StdFloat` trait the code can be made generic again relatively easy by adding a `StdFloat` bound.

**Comment by mbrubeck** [comment]
My current plan is to adopt your suggested fix, starting with ordered-float 5.0.0.

I’m waiting a bit to fix this in case there are other breaking changes we can bundle in the same release.
